unblob/handlers/compression/xz.py

```diff
@@ -1,110 +1,102 @@
-    data = bytearray(file.read(MAX_MBI_LEN))
-    file.seek(-MAX_MBI_LEN, io.SEEK_CUR)
-    size, mbi = decode_multibyte_integer(data)
-    file.seek(size, io.SEEK_CUR)
-    return size, mbi
-
-
-class XZHandler(Handler):
-    NAME = "xz"
-
-    PATTERNS = [HexString("FD 37 7A 58 5A 00")]
-
-    EXTRACTOR = Command("7z", "x", "-y", "{inpath}", "-o{outdir}")
-
-    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+def build_stream_end_scan_db(pattern_list):
+    patterns = []
+    for pattern_id, pattern in enumerate(pattern_list):
+        patterns.append(
+            (
+                pattern.as_regex(),
+                pattern_id,
+                hyperscan.HS_FLAG_SOM_LEFTMOST | hyperscan.HS_FLAG_DOTALL,
+            )
+        )
 
-        file.seek(MAGIC_BYTES_SIZE, io.SEEK_CUR)
-        stream_flags = file.read(FLAG_LEN)
-        file.seek(start_offset)
+    expressions, ids, flags = zip(*patterns)
+    db = hyperscan.Database(mode=hyperscan.HS_MODE_VECTORED)
+    db.compile(expressions=expressions, ids=ids, elements=len(patterns), flags=flags)
+    return db
 
-        for footer_offset in iterate_patterns(file, stream_flags + STREAM_FOOTER_MAGIC):
-            # make sure it's byte aligned
-            if (footer_offset - start_offset) % 4 != 0:
-                continue
 
-            end_offset = footer_offset + CRC32_LEN
+hyperscan_stream_end_magic_db = build_stream_end_scan_db(STREAM_END_MAGIC_PATTERNS)
 
-            file.seek(footer_offset - BACKWARD_SIZE_LEN)
 
-            backward_bytes = file.read(BACKWARD_SIZE_LEN)
-            stored_backward_size = convert_int32(backward_bytes, Endian.LITTLE)
-            real_backward_size = (stored_backward_size + 1) * 4
+@attr.define
+class XZSearchContext:
+    start_offset: int
+    file: File
+    end_block_offset: int
 
-            if real_backward_size > file.tell():
-                continue
 
-            file.seek(-CRC32_LEN - BACKWARD_SIZE_LEN, io.SEEK_CUR)
-            # skip backwards of backward size to the start of index
-            file.seek(-real_backward_size, io.SEEK_CUR)
+def _hyperscan_match(
+    pattern_id: int, offset: int, end: int, flags: int, context: XZSearchContext
+) -> bool:
 
-            index_size = 0
-            index_indicator = convert_int8(file.read(1), Endian.LITTLE)
-            # index indicator must be 0, per xz standard
-            if index_indicator != 0:
-                continue
+    end_offset = offset + 4
+    if end_offset < context.start_offset:
+        return True
 
-            index_size += 1
+    context.file.seek(context.start_offset, io.SEEK_SET)
+    decompressor = lzma.LZMADecompressor(format=lzma.FORMAT_XZ)
+    try:
+        size = end_offset - context.start_offset
+        for data in iterate_file(context.file, context.start_offset, size):
+            decompressor.decompress(data)
+        context.end_block_offset = end_offset
+        context.start_offset = context.file.tell()
```

Comment:
if we already matched on this footer, it is unlikely we would match something before, so no need to set start_offset

Reply:
`context.start_offset` is used when validating that the calculated size of the XZ stream corresponds to the obtained end_offset - `context.start_offset`. If we don't update the context start offset, this calculation will be off.

Example: when we match on `XZ_END_2` without setting `context.start_offset` to `XS_START_2` on the first match on `XZ_END_1`, `calculated_size != XZ_END_2 - XZ_START_1`. If we do update `context.start_offset`, then `calculated_size = XZ_END_2 - XZ_START_2`.

````
XZ_START_1
XZ_END_1
XZ_START_2
XZ_END_2
```